src/telemetry.ts

```diff
@@ -1,137 +1,200 @@
   name: string;
   version: string;
   aiKey: string;
 }
 
-
-export interface TelemetryContext {
-  properties: TelemetryProperties;
-  measurements: TelemetryMeasurements;
+/**
+ * Operation result of telemetry
+ */
+export enum TelemetryResult {
+  Succeeded = 'Succeeded',
+  Failed = 'Failed',
+  Cancelled = 'Cancelled',
 }
 
-export interface TelemetryProperties {
-  result: 'Succeeded'|'Failed'|'Cancelled';
-  error: string;
-  errorMessage: string;
-  [key: string]: string;
+/**
+ * Context of telemetry
+ */
+export interface TelemetryContext {
+  properties: {[key: string]: string};
+  measurements: {[key: string]: number};
 }
 
-export interface TelemetryMeasurements {
-  duration: number;
-  [key: string]: number;
-}
+export class TelemetryWorker {
+  private _reporter: TelemetryReporter|undefined;
+  private extensionContext: vscode.ExtensionContext|undefined;
+  private static instance: TelemetryWorker|undefined;
+
+  private constructor(context: vscode.ExtensionContext) {
+    this.extensionContext = context;
+    const extensionInfo = this.getExtensionInfo(context);
+    if (!extensionInfo) {
+      console.log('Unable to initialize telemetry');
+      return;
+    }
+    if (!extensionInfo.aiKey) {
+      console.log(
+          'Unable to initialize telemetry, please make sure AIKey is set in package.json');
+      return;
+    }
+    this._reporter = new TelemetryReporter(
+        extensionInfo.name, extensionInfo.version, extensionInfo.aiKey);
+  }
 
+  static getInstance(context: vscode.ExtensionContext): TelemetryWorker {
+    if (!TelemetryWorker.instance) {
+      TelemetryWorker.instance = new TelemetryWorker(context);
+    }
+    return TelemetryWorker.instance;
+  }
 
-function getPackageInfo(context: vscode.ExtensionContext): PackageInfo|
-    undefined {
-  const extensionPackage = require(context.asAbsolutePath('./package.json'));
-  if (extensionPackage) {
-    const packageInfo: PackageInfo = {
-      name: extensionPackage.name,
-      version: extensionPackage.version,
-      aiKey: extensionPackage.aiKey,
-    };
-    return packageInfo;
+  /**
+   * check if it is Microsoft internal user
+   */
+  private static isInternalUser(): boolean {
+    const userDomain: string = process.env.USERDNSDOMAIN ?
+        process.env.USERDNSDOMAIN.toLowerCase() :
+        '';
+    return userDomain.endsWith('microsoft.com');
   }
-  return undefined;
-}
 
-export class TelemetryWorker {
-  private static _reporter: TelemetryReporter;
+  /**
+   * Create telemetry context
+   */
+  createContext(): TelemetryContext {
+    const context: TelemetryContext = {properties: {}, measurements: {}};
+    context.properties.result = TelemetryResult.Succeeded;
+    context.properties.isInternal = TelemetryWorker.isInternalUser.toString();
+    if (this.extensionContext) {
+      context.properties.developEnvironment =
+          RemoteExtension.isRemote(this.extensionContext) ?
+          DevelopEnvironment.Container :

```

Comment:
VS Code Remote Development includes: remote SSH, WSL and container.
So we cannot infer from the remote environment that it is in a container environment.

Reply:
Nice catch. Fixed.